### ISSUE TITLE
Be able to disable/enable Clippy lints globally

### ISSUE THREAD

**Issue by repi** [EmbarkStudios/rust-ecosystem#22]
In our monorepo we want to have the same baseline of Clippy lints enabled & disabled across our crates. Right now we need to set this up manually for every crate in `lib.rs` /  `bin.rs` which is error prone and cumbersome.

This is what almost all of our crates have right now:

```rust
#![warn(clippy::all)]
#![warn(rust_2018_idioms)]
#![allow(clippy::new_ret_no_self)] // believe this is fixed in nightly https://github.com/rust-lang-nursery/rust-clippy/issues/3313
```

We would strongly prefer to instead be able to specify this in our workspace `Cargo.toml` or root `Clippy.toml`.

This is tracked in:

- [x] https://github.com/rust-lang/cargo/issues/5034
- [x] https://github.com/rust-lang/rust-clippy/issues/1313
- [x] https://github.com/rust-lang/rust-clippy/issues/3164
- [x] https://github.com/rust-lang/rust-clippy/issues/6625
- [ ] https://github.com/rust-lang/cargo/issues/12208

**Comment by YorickPeterse** [comment]
@repi I ran into this myself for [a project I am working on](https://gitlab.com/inko-lang/inko). When running Clippy you can disable lints via the CLI. I ended up putting this in an already existing Makefile, for example:

```
clippy:
	touch src/lib.rs
	${CARGO_CMD} clippy -- -Dwarnings \
		-Arenamed_and_removed_lints \
		-Aclippy::new_without_default \
		-Aclippy::needless_range_loop
```

Here the `touch src/lib.rs` is needed because otherwise Clippy does not (or at least did not in the past) build the changed files. Here the flag `-A` is used to globally disable certain lints.

Clippy also supports using a configuration file, but last I checked you can't disable lints using it.

**Comment by repi** [comment]
@YorickPeterse thanks, that does look like a workable solution for CI.

Though for working locally for users and with tools like RLS I really want a solution where everyone can just run `cargo clippy` and it  has our set of additionally enabled or disabled lints so everyone runs the same lints. Hopefully the clippy config can be extended for that. Or even better: in `Cargo.toml`

**Comment by YorickPeterse** [comment]
@repi There is a discussion about adding support for enabling/disabling lints in `Cargo.toml` in https://github.com/rust-lang/cargo/issues/5034, but it seems that at least since 2018 (example: https://github.com/rust-lang/cargo/pull/5728) not a lot of progress has been made.

**Comment by repi** [comment]
Yeah doesn't look like very active development of it unfortunately, but hope Cargo/Clippy gets it eventually.

Just added these lines on top of 30 lib.rs files in our workspace to enable more lints for our codebase.

```
#![warn(clippy::all)]
#![warn(clippy::doc_markdown)]
#![warn(clippy::dbg_macro)]
#![warn(clippy::todo)]
#![warn(clippy::empty_enum)]
#![warn(clippy::enum_glob_use)]
#![warn(clippy::pub_enum_variant_names)]
#![warn(clippy::mem_forget)]
#![warn(clippy::use_self)]
#![warn(clippy::filter_map_next)]
#![warn(clippy::needless_continue)]
#![warn(clippy::needless_borrow)]
#![warn(rust_2018_idioms)]
```

Copy'n'paste is not pretty, but works.

**Comment by repi** [comment]
This is getting a bit more silly without this support, we have like 30 crates in our repo where all of their `lib.rs` start something like this:

```rust
#![forbid(unsafe_code)]
#![warn(
    clippy::all,
    clippy::doc_markdown,
    clippy::dbg_macro,
    clippy::todo,
    clippy::empty_enum,
    clippy::enum_glob_use,
    clippy::pub_enum_variant_names,
    clippy::mem_forget,
    clippy::use_self,
    clippy::filter_map_next,
    clippy::needless_continue,
    clippy::needless_borrow,
    clippy::match_wildcard_for_single_variants,
    clippy::if_let_mutex,
    clippy::mismatched_target_os,
    clippy::await_holding_lock,
    clippy::match_on_vec_items,
    clippy::imprecise_flops,
    clippy::suboptimal_flops,
    clippy::lossy_float_literal,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::fn_params_excessive_bools,
    clippy::exit,
    clippy::inefficient_to_string,
    clippy::linkedlist,
    clippy::macro_use_imports,
    clippy::option_option,
    clippy::verbose_file_reads,
    clippy::unnested_or_patterns,
    rust_2018_idioms,
    future_incompatible,
    nonstandard_style
)]
```

Does work though!

**Comment by repi** [comment]
We now have ~170 crates across ~20 repositories so have an even larger list of Clippy lints we enable by default across them now scattered and duplicated across 170 locations, which makes it hard to add and esp easy to miss when creating new crates. 

We really do need the proper Cargo/Clippy solution for a `clippy.toml` where we can specify the above list of warn/deny/forbid lints that can be applied for the entire repository: https://github.com/rust-lang/cargo/issues/5034. We'll try and investigate if we can help push the development of that and clippy further. It is a fantastic tool that we want to use more of!

**Comment by repi** [comment]
We had a good sync with the Clippy team a couple of weeks ago and this is one of the top issues in the Clippy [2021 Roadmap](https://github.com/rust-lang/rust-clippy/pull/6462)!

**Comment by cztomsik** [comment]
omg yes!

**Comment by jplatte** [comment]
We just hit a very weird issue with the `.cargo/config.toml`-based workaround, where `cargo <command>` and `cargo <command> --target foo` would partially invalidate each other's caches. Turned out a historical design decision in Cargo was the culprit: https://github.com/rust-lang/cargo/issues/4423

Luckily, there is a workaround, you just add this to the `config.toml` to fix the problem on Nightly (I don't know of any solutions for stable):

```toml
# outside any section
target-applies-to-host = false

[unstable]
target-applies-to-host = true
```

PR that fixes this for our project, with extra comments: https://github.com/matrix-org/matrix-rust-sdk/pull/737/files

**Comment by jplatte** [comment]
Regarding my previous comment, do not use that nightly feature if you also want to pass `RUSTDOCFLAGS` to rustdoc through `cargo +nightly doc`, the flags will silently get ignored: https://github.com/rust-lang/cargo/issues/10744

**Comment by jplatte** [comment]
One more update: I've found a good-ish workaround: Just reset target-applies-to-host to its default value when building docs, for example by setting the environment variable `CARGO_TARGET_APPLIES_TO_HOST` to `true`. (cc https://github.com/matrix-org/matrix-rust-sdk/pull/757)

**Comment by 9SMTM6** [comment]
Btw, I've found a different (kinda horrible) workaround:

You can add another binary that declares all the other entrypoints to be linted as modules, and then add the lints in that binary.

It seems to have a number of drawbacks that you'll probably find out quickly (ie it triggers dead_code), but I just thought I'd leave that here, perhaps it turns out to be the least evil for some and/or the main problems can be fixed.

**Comment by repi** [comment]
Ah thanks for sharing! Kinda horrible indeed but good to know about!

**Comment by jplatte** [comment]
Ugh, I found that `target-applies-to-host` *entirely* disables rustflags-based clippy configuration (it's not just rustdoc) on nightly for regular builds without an explicit `--target`. This feature breaks more things than it fixes 🙄

**Comment by ericseppanen** [comment]
I wrote [cargo-cranky](https://github.com/ericseppanen/cargo-cranky), which allows easy clippy lint configuration. Lint config goes in a `Cranky.toml` file and then `cargo cranky` just runs `cargo clippy` with the right command line that enables those lints. It's pretty basic so far, but I find it a nice improvement, as it can handle every bin/lib/example in a workspace.

**Comment by avantgardnerio** [comment]
I need this for the `single_range_in_vec_init` rule, which is telling me valid code is invalid and offering fixes that would alter behavior:

```
warning: a `Vec` of `Range` that is only one element
   --> datafusion/physical-expr/src/window/rank.rs:291:32
    |
291 |         test_f64_result(&r, 7, vec![0..7], expected)?;
    |                                ^^^^^^^^^^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#single_range_in_vec_init
help: if you wanted a `Vec` that contains the entire range, try
    |
291 |         test_f64_result(&r, 7, (0..7).collect::<std::vec::Vec<usize>>(), expected)?;
    |                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: if you wanted a `Vec` of len 7, try
    |
291 |         test_f64_result(&r, 7, vec![0; 7], expected)?;
    |                                     ~~~~
```
